compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -0,0 +1,347 @@
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
+
+    private val getOrCreateKotlinClass =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinClass") }
+
+    private val getOrCreateKotlinPackage =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinPackage") }
+
+    private val arrayItemGetter =
+        context.ir.symbols.array.owner.functions.single { it.name == Name.identifier("get") }
+
+    private val kPropertyType =
+        context.reflectionTypes.kProperty.toIrType()
+
+    private val kPropertiesFieldType =
+        context.ir.symbols.array.createType(false, listOf(makeTypeProjection(kPropertyType, Variance.OUT_VARIANCE)))
+
+    // TODO: remove code duplication with CallableReferenceLowering
+    private val IrMemberAccessExpression.parentJavaClassReference
+        get() = IrClassReferenceImpl(
+            startOffset, endOffset,
+            plainJavaClass.defaultType,
+            plainJavaClass.symbol,
+            CrIrType(context.state.typeMapper.mapImplementationOwner(descriptor))
+        )
+
+    private fun IrBuilderWithScope.buildReflectedContainerReference(expression: IrMemberAccessExpression): IrExpression {
+        val parent = expression.getter?.owner?.parent ?: expression.field?.owner?.parent
+        return if (parent is IrPackageFragment) {
+            irCall(getOrCreateKotlinPackage as IrFunction).apply {
+                putValueArgument(0, expression.parentJavaClassReference)
+                putValueArgument(1, irString(this@PropertyReferenceLowering.context.state.moduleName))
+            }
+        } else if (parent is IrClass) {
+            irCall(getOrCreateKotlinClass as IrFunction).apply {
+                putValueArgument(0, expression.parentJavaClassReference)
+            }
+        } else {
+            irNull()
+        }
+    }
+
+    private class PropertyReferenceKind(
+        val interfaceSymbol: IrClassSymbol,
+        val reflectedSymbol: IrClassSymbol,
+        val wrapper: IrFunction
+    )
+
+    private fun propertyReferenceKind(mutable: Boolean, i: Int) = PropertyReferenceKind(
+        context.getIrClass(FqName("kotlin.jvm.internal.${if (mutable) "Mutable" else ""}PropertyReference${i}")),
+        context.getIrClass(FqName("kotlin.jvm.internal.${if (mutable) "Mutable" else ""}PropertyReference${i}Impl")),
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier(if (mutable) "mutableProperty$i" else "property$i") }
+                as IrFunction
+    )
+
+    private fun propertyReferenceKindFor(expression: IrMemberAccessExpression, scope: Scope): PropertyReferenceKind =
+        expression.getter?.owner?.let {
+            val boundReceivers = listOf(expression.dispatchReceiver, expression.extensionReceiver).count { it != null }
```

Comment:
`listOfNotNull(...).size` here and below (will get rid of the IDE inspection about `it`)